adapter/src/main/java/com/twilio/kudu/dataloader/DataLoader.java

```diff
@@ -263,67 +276,95 @@ private void bindValues(PreparedStatement statement, UniformLongValueGenerator t
-        // subset
-        // of time ranges
-        UniformLongValueGenerator subsetTimestampGenerator = new UniformLongValueGenerator(finalStartTimestamp,
-            finalEndTimestamp);
         try (Connection conn = DriverManager.getConnection(url)) {
           // Create prepared statement that can be reused
           String sql = buildSql();
           PreparedStatement stmt = conn.prepareStatement(sql);
-
           // populate table with data
-          for (int i = 1; i <= numRowsWrittenPerThread; ++i) {
-            bindValues(stmt, subsetTimestampGenerator);
-            stmt.execute();
-            if (i % batchSize == 0) {
-              conn.commit();
-              logger.info("Total number of rows committed {} time taken for " + "current batch {}", i,
-                  (System.currentTimeMillis() - threadStartTime));
-              threadStartTime = System.currentTimeMillis();
+          int rowCount = 0;
+          long batchStartTimestamp;
+          long batchEndTimestamp = threadStartTimestamp;
+          for (int i = 1; i <= numCubeMutationBatches; ++i) {
+            long rangePerBatch = (threadEndTimestamp - threadStartTimestamp) / numCubeMutationBatches;
+            long cubeDelta = Math.max(rangePerBatch, cubeGranularityFloorMod);
+            batchStartTimestamp = batchEndTimestamp;
+            batchEndTimestamp = i == numCubeMutationBatches ? threadEndTimestamp
+                : SqlFunctions.floor(batchStartTimestamp + cubeDelta, cubeGranularityFloorMod);
+            logger.info("batch {} start timestamp {}", i, new Date(batchStartTimestamp));
+            logger.info("batch {} end timestamp {}", i, new Date(batchEndTimestamp));
+            // create a timestamp column value generator that generates timestamps for this
+            // subset of time ranges
+            UniformLongValueGenerator subsetTimestampGenerator = new UniformLongValueGenerator(batchStartTimestamp,
+                batchEndTimestamp);
+            int numRowsInBatch = numRowsPerThread / numCubeMutationBatches;
+            // add any remaining rows to the last batch
+            if (i == numCubeMutationBatches) {
+              numRowsInBatch += numRowsPerThread % numCubeMutationBatches;
+            }
+            for (int j = 1; j <= numRowsInBatch; ++j) {
+              bindValues(stmt, subsetTimestampGenerator);
+              stmt.execute();
+              if (++rowCount % COMMIT_BATCH_SIZE == 0) {
+                conn.commit();
+                logger.info("Total number of rows committed {} time taken {}", rowCount,
+                    (System.currentTimeMillis() - threadStartTime));
+                threadStartTime = System.currentTimeMillis();
+              }
             }
+            // commit any remaining rows
+            conn.commit();
+            KuduMetaImpl kuduMetaImpl = (conn.unwrap(KuduCalciteConnectionImpl.class)).getMeta();
+            kuduMetaImpl.clearMutationState();
```

Comment:
Once we are done with this subset of the time range we can clear the cube mutation state thus freeing up memory.

Reply:
💯